### PR TITLE
[SofaCore] Move the streaming operator from .h to .cpp

### PR DESCRIPTION
**Pull request by damienmarchal**
There is no need to have the operator<<() fully defined in the class declaration.
A better code decoupling can be obtained in the following way.

Let's suppose we want to support a new class named MyClass. 
A more opaque design is composed of three parts.

1) declare the function << for a given type, (eg: in the vicinity of  the forward declaration of MyClass in fwd.h):
```cpp 
SOFA_CORE_API std::ostream& operator<<(..., MyClass&);
```

2) define the function in a MyClass.cpp (or anywhere you prefer):
```cpp 
operator<<(..., MyClass& a){ ....put the implementation here... }
```

3) This one is optional only if the implementation of the operator requires access to private data field). If this happens then you need an exra step to specify the following 
```cpp 
class MyClass { friend operator<<(..., MyClass&); }
``` 

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
This one is passing the CI, it is pretty hamless so please review pass it to ready if you agree :)

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests][with-regression-tests]

File: SofaKernel/modules/SofaBaseTopology/CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,7 @@ set(SOFABASETOPOLOGY_SRC "src/${PROJECT_NAME}")
 
 set(HEADER_FILES
     ${SOFABASETOPOLOGY_SRC}/config.h.in
+    ${SOFABASETOPOLOGY_SRC}/fwd.h
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/CMakeLists.txt:8]: I'm not very fan of this header name... especially if you have several fwd.h

File: SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp [hunk 2]
```diff
@@ -1207,4 +1207,59 @@ void TetrahedronSetTopologyContainer::updateTopologyEngineGraph()
     TriangleSetTopologyContainer::updateTopologyEngineGraph();
 }
 
+std::ostream& operator<< (std::ostream& out, const TetrahedronSetTopologyContainer& t)
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp:1210]: I'm not sure this make really sens to have iostream operator for a TopologyContainer.
I would be curious to see where it is used.

Comment (epernod) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp:1210]: moreover it is not using all the "modifier" methods (AddTetrahedron, etc.) that handle change in the topology and propagate the events

Comment (damienmarchal) [reply SofaKernel/modules/SofaBaseTopology/src/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp:1210]: I didn't  wanted to change the existing code... just move the stream out of the .h. Refactoring this part of sofa code would be very nice but it is not my topic right now and I will not pick a new one until I finished the one I'm in ;)

File: SofaKernel/modules/SofaBaseTopology/CMakeLists.txt (hunk 1 above)
Comment (damienmarchal) [reply SofaKernel/modules/SofaBaseTopology/CMakeLists.txt:8]: Me neither but as we have agreed for having multiple config.h I think it is better to be consistent.

**Comment by damienmarchal** [comment]
I fixed the confict :)

**Comment by damienmarchal** [comment]
@hugtalbot  why breaking ?

**Pull request merged**